x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/strip_non_ecs_fields.ts

```diff
@@ -0,0 +1,209 @@
+      const isEcsField = getIsEcsField(pathToValidate);
+      const isEcsFieldObject = getIsEcsFieldObject(pathToValidate);
+
+      // if field is in Ecs mapping and not object, the whole path is invalid
+      if (isEcsField && !isEcsFieldObject) {
+        isValid = false;
+      }
+      return pathToValidate;
+    }, '');
+
+  return isValid;
+};
+
+/**
+ * check whether source field value is ECS compliant
+ */
+const computeIsEcsCompliant = (value: SourceField, path: string) => {
+  // if path consists of dot-notation, ensure each path within it is ECS compliant (object or flattened)
+  if (path.includes('.') && !validateDottedPathInEcsMappings(path)) {
+    return false;
+  }
+
+  const isEcsField = getIsEcsField(path);
+
+  // if field is not present is ECS mapping, it's valid as doesn't have any conflicts with existing mapping
+  if (!isEcsField) {
+    return true;
+  }
+
+  const ecsField = ecsFieldMap[path as keyof typeof ecsFieldMap];
+  const isEcsFieldObject = getIsEcsFieldObject(path);
+
+  // validate if value is a long type
+  if (ecsField?.type === 'long') {
+    return isValidLongType(value);
+  }
+
+  // validate if value is a numeric type
+  if (ecsField?.type === 'float' || ecsField?.type === 'scaled_float') {
+    return isValidNumericType(value);
+  }
+
+  // validate if value is a valid ip type
+  if (ecsField?.type === 'ip') {
+    return isValidIpType(value);
+  }
+
+  // validate if value is a valid date
+  if (ecsField?.type === 'date') {
+    return isValidDateType(value);
+  }
+
+  // validate if value is a valid boolean
+  if (ecsField?.type === 'boolean') {
+    return isValidBooleanType(value);
+  }
+
+  // if ECS mapping is JS object and source value also JS object then they are compliant
+  // otherwise not
+  return isEcsFieldObject ? isPlainObject(value) : !isPlainObject(value);
+};
+
+interface StripNonEcsFieldsReturn {
+  result: SourceFieldRecord;
+  removed: Array<{ key: string; value: SearchTypes }>;
+}
+
+/**
+ * strips alert source object from ECS non compliant fields
+ */
+export const stripNonEcsFields = (doc: SourceFieldRecord): StripNonEcsFieldsReturn => {
+  const result = cloneDeep(doc);
+  const removed: Array<{ key: string; value: SearchTypes }> = [];
+
+  /**
+   * traverses through object and deletes ECS non compliant fields
+   * @param document - document to traverse
+   * @param documentKey - document key in parent document, if exists
+   * @param parent - parent of traversing document
+   * @param parentPath - path of parent in initial source document
+   */
+  const traverseAndDeleteInObj = (
+    document: SourceField,
+    documentKey: string,
+    parent?: SourceFieldRecord,
+    parentPath?: string
+  ) => {
+    const fullPath = [parentPath, documentKey].filter(Boolean).join('.');
+    // if document array, traverse through each item w/o changing documentKey, parent, parentPath
+    if (isArray(document)) {
+      document.slice().forEach((value) => {
```

Comment:
`traverseAndDeleteInObj` function which is called for each item in array, can remove this item, if it's not compliant.

This shifts iteration of `forEach` loop and some of items can be missed because `document` is mutated and loses non-compliant item. So, I clone document array using `slice()`, thus ensuring every item is traversed.